flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricQueryService.java

```diff
@@ -0,0 +1,180 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.runtime.metrics;
+
+import akka.actor.ActorRef;
+import akka.actor.ActorSystem;
+import akka.actor.Props;
+import akka.actor.UntypedActor;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.metrics.CharacterFilter;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.Histogram;
+import org.apache.flink.metrics.HistogramStatistics;
+import org.apache.flink.metrics.Metric;
+import org.apache.flink.runtime.metrics.groups.AbstractMetricGroup;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The MetricQueryService creates a key-value representation of all metrics currently registered with Flink when queried.
+ *
+ * It is realized as an actor and receives the following messages:
+ * - {@code Tuple3<String, Metric, AbstractMetricGroup> => Notification of added Metric}
+ * - {@code Metric => Notification of removed Metric}
+ * - {@code ActorRef => Query for metric dump}
+ */
+public class MetricQueryService extends UntypedActor {
+	private static final Logger LOG = LoggerFactory.getLogger(MetricQueryService.class);
+
+	public static final byte CATEGORY_COUNTER = 0;
+	public static final byte CATEGORY_GAUGE = 1;
+	public static final byte CATEGORY_HISTOGRAM = 2;
+
+	private final Map<Gauge<?>, String> gauges = new HashMap<>();
+	private final Map<Counter, String> counters = new HashMap<>();
+	private final Map<Histogram, String> histograms = new HashMap<>();
+
+	private final List<Metric> toRemove = new ArrayList<>();
+
+	@Override
+	public void onReceive(Object message) throws Exception {
+		try {
+			if (message instanceof Tuple3) { // add metric
+				Tuple3<String, Metric, AbstractMetricGroup> tuple = (Tuple3<String, Metric, AbstractMetricGroup>) message;
+
+				String metricName = tuple.f0;
+				Metric metric = tuple.f1;
+				AbstractMetricGroup group = tuple.f2;
+
+				String name = group.getQueryServiceMetricIdentifier(metricName, new CharacterFilter() {
+					@Override
+					public String filterCharacters(String input) {
+						return input.replaceAll("[ :.,]", "_");
+					}
+				});
+
+				if (metric instanceof Counter) {
+					counters.put((Counter) metric, name);
+				} else if (metric instanceof Gauge) {
+					gauges.put((Gauge<?>) metric, name);
+				} else if (metric instanceof Histogram) {
+					histograms.put((Histogram) metric, name);
+				}
+			} else if (message instanceof Metric) { // remove metric
```

Comment:
The same applies to the remove metric message and the create dump. Both should get a proper message, e.g. `RemoveMetric` and `CreateDump`.